haystack/utils/matching.py

```diff
@@ -0,0 +1,127 @@
+from collections import namedtuple
+from typing import Generator, Tuple, List, Union
+import re
+from rapidfuzz import fuzz
+from multiprocessing import Pool
+from tqdm import tqdm
+from itertools import groupby
+
+
+_CandidateScore = namedtuple("_CandidateScore", ["context_id", "candidate_id", "score"])
+
+
+def _score_candidate(args: Tuple[Union[str, Tuple[object, str]], Tuple[object, str], int]):
+    context, candidate, min_words = args
+    candidate_id, candidate_text = candidate
+    context_id, context_text = (None, context) if isinstance(context, str) else context
+    score = calculate_context_similarity(context_text, candidate_text, min_words)
+    return _CandidateScore(context_id=context_id, candidate_id=candidate_id, score=score)
+
+
+def normalize_white_space_and_case(str):
+    return re.sub(r"\s+", " ", str).lower().strip()
+
+
+def calculate_context_similarity(context: str, candidate: str, min_words: int = 25) -> float:
+    """
+    Calculates the text similarity score of context and candidate.
+    The score's value ranges between 0.0 and 100.0.
+
+    :param context: The context to match.
+    :param candidate: The candidate to match the context.
+    :param min_words: The minimum number of words context and candidate need to have in order to be scored.
```

Comment:
`words` should be generally better understandable then `tokens` by most people. Also `min_seq_len` refers to a sequence of "whatever" because it totally depends on the tokenizer what you're dealing with (words, wordpieces, bytes, etc.). Here we're actually dealing with words. So I'd leave it like that.